libraries/AP_OpenDroneID/AP_OpenDroneID.cpp

```diff
@@ -0,0 +1,759 @@
+        _classification_type = classification_type;
+    }
+    else {
+        _classification_type = MAV_ODID_CLASSIFICATION_TYPE_UNDECLARED;
+    }
+}
+
+void AP_OpenDroneID::set_category_eu(MAV_ODID_CATEGORY_EU category_eu)
+{
+    //check if category_eu is valid
+    if ((category_eu == MAV_ODID_CATEGORY_EU_OPEN) || (category_eu == MAV_ODID_CATEGORY_EU_SPECIFIC)
+        || (category_eu == MAV_ODID_CATEGORY_EU_CERTIFIED)
+    ) {
+        _category_eu = category_eu;
+    }
+    else {
+        _category_eu = MAV_ODID_CATEGORY_EU_UNDECLARED;
+    }
+}
+
+void AP_OpenDroneID::set_class_eu(MAV_ODID_CLASS_EU class_eu)
+{
+    //check if class_eu is valid
+    if ((class_eu == MAV_ODID_CLASS_EU_CLASS_0) || (class_eu == MAV_ODID_CLASS_EU_CLASS_1) || (class_eu == MAV_ODID_CLASS_EU_CLASS_2)
+        || (class_eu == MAV_ODID_CLASS_EU_CLASS_3) || (class_eu == MAV_ODID_CLASS_EU_CLASS_4) || (class_eu == MAV_ODID_CLASS_EU_CLASS_5)
+        || (class_eu == MAV_ODID_CLASS_EU_CLASS_6) ) {
+        _class_eu = class_eu;
+    }
+    else {
+        _class_eu = MAV_ODID_CLASS_EU_UNDECLARED;
+    }
+}
+void AP_OpenDroneID::set_operator_id_type(MAV_ODID_OPERATOR_ID_TYPE type)
+{
+    //check if type is valid
+    if (type == MAV_ODID_OPERATOR_ID_TYPE_CAA) {
+        _operator_id_type = type;
+    }
+    else {
+        _operator_id_type = MAV_ODID_OPERATOR_ID_TYPE_CAA;
+    }
+}
+
+void AP_OpenDroneID::set_operator_id(char* id)
+{
+    memcpy(_operator_id,id,20);
+}
+
+void AP_OpenDroneID::set_description_type(MAV_ODID_DESC_TYPE type)
+{
+    //check if type is type
+    if (type == MAV_ODID_DESC_TYPE_TEXT) {
+        _description_type = type;
+    }
+    else {
+        _description_type = MAV_ODID_DESC_TYPE_TEXT;
+    }
+}
+
+void AP_OpenDroneID::set_description(char* description)
+{
+    memcpy(_description,description,23);
+}
+
+void AP_OpenDroneID::set_operator_latitude(float latitude)
+{
+    _operator_latitude = latitude;
+}
+
+void AP_OpenDroneID::set_operator_longitude(float longitude)
+{
+    _operator_longitude = longitude;
+}
+
+void AP_OpenDroneID::set_operator_timestamp(time_t epoch_timestamp)
+{
+    time_t timestamp_2019 = epoch_timestamp - (time_t) 1546300800; //1546300800 = 1/1/2019 00:00:00 UTC
+    if (timestamp_2019 > 0) {
+        _operator_timestamp = (uint32_t) timestamp_2019;
+    }
+    else {
+        _operator_timestamp = 0;
+    }
+}
+
+void AP_OpenDroneID::set_area_count(uint16_t count)
+{
+    if (count > 65000) { //according the standard maximum value = 65000
+        count = 65000;
+    }
+
+    _area_count = count;
+}
+
+void AP_OpenDroneID::set_area_radius(uint16_t radius)
+{
+    if (radius > MAX_AREA_RADIUS) {
+        radius = MAX_AREA_RADIUS;
+    }
+    _area_radius = radius;
+}
+
+void AP_OpenDroneID::set_area_ceiling(float ceiling)
+{
+    if (ceiling < MIN_ALT) {
+        ceiling = MIN_ALT;
+    }
+    else if (ceiling > MAX_ALT) {
+    }
+    _area_ceiling = ceiling;
```

Comment:
```suggestion
    _area_ceiling = constrain_float(ceiling, ODID_MIN_ALT, ODID_MAX_ALT;
```

Reply:
in this new revision I used contrain functions as much as possible.